### ISSUE TITLE
GUI: initial screen update is just wrong!

### ISSUE THREAD

**Issue by stacksmith** [neovim/neovim#7723]
NVIM v0.2.2  xubuntu 17.04 zesty  no terminal/messagepack only.

# Problem description:

Initial screen grid update is sent from nvim _before_ modes are established and default styles and colors are set.  The screen is cleared before colors are established.  Most of the screen, with the exception of the first line, is then filled with bold spaces, before styling information is established.   Status line is displayed as inverted, once again, before the colors are established.

After mode styling information is transmitted, no refresh takes place.

# Expected behaviour:

1) establish ground rules - color, styles, modes, etc.
2) enter correct state
3) clear the screen
4) send the text updates for areas that are changed _only_.

Should any text, color, style or mode data changed, 

5) send text updates for areas that are affected _only_.

# Actual behaviour:

1) text updates sent before colors are established
2) first line is strangely different from the rest of the screen on startup (bold spaces!  While you think it does not matter, bold styling in a GUI environment may be drastically different visually from non-bold!)
3) monumental changes to colors, styles and mode defaults are not followed by a grid update requiring guesswork or a deep understanding of sequencing of NVIM events (impossible with current quality of documentation, although it specifically says not to expect anything to be in sequence - a nonsensical pronouncement for a protocol that demands sequencing)
4) redundant commands are sent

And finally, forgive me for beating a dead horse, I would like to again point out the not just the initial sequencing is wrong, but pretty much all of messagepack data is so **poorly designed, badly sequenced, inconsistent, and largely undocumented**  that I strongly suspect it was deliberately built as a torture instrument.

**Comment by justinmk** [comment]
You keep mentioning "bad sequencing", maybe you have experience in protocol design, event systems, etc.? Help is welcome. Please help. 

The current design is just the current design. A lot of work has gone into implementing more structured UI events. And more work could be done to eliminate redundancy or whatever.

Have a look at the issues tab up there ^  ... the problems you mention with exclamation marks just aren't causing major problems at the moment.

**Comment by justinmk** [comment]
> or a deep understanding of sequencing of NVIM events (impossible with current quality of documentation, although it specifically says not to expect anything to be in sequence - a nonsensical pronouncement for a protocol that demands sequencing)

Can you clarify this? There are ~16 UI implemenations, I'm certain most of those authors don't have a deep understanding of "sequencing of nvim events".

> although it specifically says not to expect anything to be in sequence

Which part of the help are you referring to exactly? Events don't arrive "out of order". Only they aren't guaranteed to happen at the edges of a synchronous RPC request, for example: they may arrive in the middle of an RPC request/response pair.

**Comment by stacksmith** [comment]
I understand that it is difficult and would be happy to help.  Unfortunately, now that the protocol is in place and many GUI systems are using it, change requires even more rigorous thinking, of a kind that has proved itself elusive.

In the meantime, rather than defensive arguing about where in the manual it says not to expect any messagepack data to be in any sequence, I again point to the problem I described above.  While what I described is certainly not a deadly bug, it is entirely inconsistent and requires guesswork - or works only if the color and style defaults just happen to be in-line with nvim's ideas of what a GUI should look like, which is again guesswork.

For instance, should the user interpret 'bold' highlight as a different color background, the initial screen looks weird with the first line not cleared.

**Comment by justinmk** [comment]
> rather than defensive arguing about where in the manual it says not to expect any messagepack data to be in any sequence,

I asked about the manual so that I can clarify it, correct the doc, or adjust my own understanding.

>  Unfortunately, now that the protocol is in place and many GUI systems are using it, change requires even more rigorous thinking, of a kind that has proved itself elusive.

UIs can opt-in to new events (`:help ui-ext-options`), and we plan to offer events that represent the screen with more granularity and less terminal/Vim legacy. So, it's not too late, and is never too late with the API extensibility model. Rigor notwithstanding.

**Comment by stacksmith** [comment]
All right, @Justmink, please accept my apologies as it is certainly not my goal to offend you.  I am obviously having one of those days in addition to an autism spectrum flareup.  I am a neovim fan, and am trying to help, and appreciate both the breadth of the project and your part in it.  Having said that, the protocol has caused me some pain and is extremely flawed (somewhat understating the harsh reality of the last few days of hacking).  

I direct you to the documentation at https://neovim.io/doc/user/msgpack_rpc.html, the bullet that states
```
- Don't assume anything about the order that responses to msgpack-rpc requests will arrive.
```
This is entirely counterintuitive as sequencing is implied in the task at hand.  Setting colors before drawing text is very different from reverse of that.

Opt-in events will certainly help, but too much of that leads to a complete version hell.  

And again, what I am bringing up here, is a sequencing issue that can be solved without protocol changes.

**Comment by bfredl** [comment]
> - Don't assume anything about the order that responses to msgpack-rpc requests will arrive.

I think was meant from the perspective of implementing a msgpack-rpc event loop, in which there is no situation in which it can be  assumed some specific message will arrive next, because many independent sources can always cause requests and notifications. It doesn't mean that events from the same source will be disordered, like two `rpcnotify()` calls ordered in the same function. Some common sense has to be used: if a drawing API is stateful (as low-level drawing APIs tend to be) one has to assume that events are ordered at least to the extent for this state to make sense.

> text updates sent before colors are established

Colors and styles are never "established", rather communicated just before each specific "put". I think you rather see some screen updates before the user's color scheme has been loaded. We can never completely avoid this as nvim might need to communicate with the user before their vimrc is fully loaded.

> colors, styles and mode defaults are not followed by a grid update

I'm not sure what this means. You will interleavedly see "highlight_set" and "put" commands. Each "put" command uses the style from the last "highlight_set". This is not different from many traditional drawing APIs (with a stateful "pen" or "drawing context" or whatever). Modes are independent of grid which is documented in the last ui.txt revision.

>  I strongly suspect it was deliberately built as a torture instrument.

Rather it started (and still to much extent is) as glorified terminal emulator layer. And then it has gradually gained GUI-specific features. It might be somewhat of a incoherent mess, but this way we can actually roll out UI features to users, instead of spending two years implementing a "well designed" UI from scratch, with no useful updates to users.
Just redesigning the msgpack wire format might actually be a realistic goal, the thing is just that it seems to work "well enough" for so many existing UI implementations (including some pretty advanced ones, such as embeddings in other editors) so no one has really bothered with this project.

**Comment by stacksmith** [comment]
@bredl, thanks for clarifying.  I would like to caution everyone about stating that so many UI implementations are fine as is.  Mine will be a fine implementation when I am done, but instead of a day project it will be a month-long.  A well-working embedding support would guarantee 'vim everywhere' (as much as vim is acceptable to people anyway). 

This is a fairly common situation which leads to problems - consider a restaurant owner who thinks his business is doing fine while the customers just don't bother telling him that his food tastes awful, and after eating once and getting as far away as possible never come back.  I held my tongue for a long time, thinking that I must be doing something wrong because others seem to have used the GUI interface.  And inevitably, you wind up being hated for saying what needs to be said.

Since there are some ad-hoc semantic protocols added, I think that it's worth to consider that there some distinct layers here.   Fundamentally, the GUI should not ever have to guess what neovim is up to or will do: this information should come from neovim.

* A screen updater
It should simply send grid updates so that a front-end can draw the screen in a reasonable way.  Modern antialiased font engines overlap glyphs even when using monospaced fonts, so even this needs rethinking as just blitting characters to the right positions on the screen does not work.  Perhaps style information - which is currently 5 bits (bold, inverted, etc) plus up to 3 color indices should be just attached to characters, and entire 'dirty line' transmitted in one shot.  Otherwise the GUI has to do that, which makes implementation that much harder.    I wound up having to maintain my own grid with 32 bits per character (21 utf8 bits, 11 'style' bits), a style mapper, a color mapper and a bunch of other infrastructure just to be able to draw nice-looking monospaced characters on the screen with enough headroom to draw lots of colors.  (Again, working well enough is a relative term, when your API users are literally bending over backwards to make the API work).

* Skin layer
The GUI should be in charge of what the screen looks like.  While neovim - the standalone application - runs the show, neovim the embedded editor should definitely not even have the concept of RGB colors.  However just color indices is not enough - semantic names would be more helpful.

* A semantic layer
This layer should ##ASK THE GUI## if it is OK to perform commands.  There are many ways to do things in VIM, and the GUI should not have to filter character sequences to be able to jail neovim to do specific tasks which may be as small as entering a password or a text field.  In many cases it is inappropriate for an embedded editor to open files or start a terminal mode.  Alternatively, a command map may communicated by the gui to restrict certain classes of operation.

* File protocol
As far as files go, an embedded neovim should certainly go through the GUI - which may run on a different machine - to get file data.

These are just a few thoughts, after spending a few weeks implementing a Common Lisp neovim wrapper from scratch, including a non-allocating messagepack layer and an XCB wrapper.  Spending time underneath a car gives you a different view than that of the engineer who designed the shiny parts.  I think it is great that neovim has an embedding capability, but please don't let it go to waste because 'so many existing implementations are fine as is'.

**Comment by stacksmith** [comment]
By the way, loading vimrc without asking the GUI is not a good idea.  The gui should probably send a command to neovim to load vimrc, or some other file.

**Comment by justinmk** [comment]
> This layer should ASK THE GUI if it is OK to perform commands

How will that work with multiple attached UIs?

> but please don't let it go to waste because 'so many existing implementations are fine as is'.

Your comments are appreciated. But also I don't know what to say, because we have spent many days improving the API. We only have N days. We used N days to do the things that were done so far. There were no days left to do more things. We will do more things in the future.

**Comment by stacksmith** [comment]
@justinmk - I am assuming that each UI 'attachment' gets its own context.  Can you currently attach multiple UIs to the same instance of neovim?  My experiments simply start a new process and use stdin/stdout to communicate.  Is it possible to connect more than oce using sockets?

Anyway, each connection should get an entirely independent state, of course.

I totally understand the time/resource constraints.  In the ideal world, of course, neovim is nothing but a server with different GUIs attached to it (doing that would make RPCs the main mechanism, not just an afterthought - and guarantee their correctness and efficiency).  But it is what it is.

I would love to help when you have time to work on protocols-related business, as I have a painful understanding of some issues that are not immediately obvious.

**Comment by justinmk** [comment]
> Can you currently attach multiple UIs to the same instance of neovim?
> ...
> Is it possible to connect more than once using sockets?

yes and yes. 

However just to note, startup/attachment-time for UIs _is_ a big pain point and one of the next ~5 things I personally plan to work on.


Also I should have added, the point of my last comment is to emphasize that complacency is _not_ a risk, as far as I am concerned.

**Comment by stacksmith** [comment]
@justinmk - that is great, but can you clarify - if I attach for the second time, do I just get the same grid as the first attachment - or do I get a clean slate?  There are important uses for both cases, and each would benefit from not starting another process.

Just to be clear - I am not accusing you or the neovim team of complacency - god knows you've been productive enough.  As someone who has been focused on a small part of this puzzle - I can see how it could have been done in a slightly different way that would make GUI client implementations really simple and pretty much stateless.  Hence my frustrations.  And I totally understand that you did not have enough time to do it right - which is pretty much impossible until you try to actually write a GUI and see for yourself what the real issues are.  Even then, there are different and distinct use cases for an embedded neovim that probably have different requirements.

**Comment by bfredl** [comment]
> I would like to caution everyone about this sentiment that things are fine because so many UI implementations are fine as is. [...] I think it is great that neovim has an embedding capability, but please don't let it go to waste because 'so many existing implementations are fine as is'.

I never said "everything is fine as it is". Rather, we more or less started with the abstraction (or at least "separation") that vim had beneath TUI and GUI, and has worked (not much yet, until very recently) to improve it from there. The limited improvements has already made some people happy. Further improvements will surely make even more people happy, possibly including yourself. But even if that doesn't happen absolutely nothing is "wasted", because the existing stable work has already proven very useful to quite a number people.

>  Modern systems often require each color to be registered (or a pen created). Just sending an RGB value requires the front end to maintain a color map and do translation and lookup. Looking up a local equivalent of a 24-bit RGB value requires something like a hashtable. [..] This idea of client handles is highly underrated. 

These  are good things to note. As it happens, color info is _already_ abstracted between two formats (255 palette and rgb), so a third format that instead uses allocated color handles could be added without too much friction. We already use "client handles" for a lot of things, a PR for color handles would be very welcome.

> I must be doing something wrong because others seem to have used the GUI interface

It might be that others has looked into existing ui:s and used them as a reference implemtation, rather than the documentation that admittedly sucks. I know this for a fact because (1) I wrote the docs, under the theory that bad docs are better than none and people capable of writing good technical documentation unlike me would come and improve it and (2) many of the successful UIs are older than than any text that now is in `ui.txt`.

> A semantic layer [..] File protocol

All these might be good ideas, but they are separate concerns from the UI protocol, and best discussed in separate issues.

**Comment by bfredl** [comment]
> do I just get the same grid as the first attachment - or do I get a clean slate?

As it happens I'm looking into supporting multiple grids. It will be a while until (if ever) we can give different clients "a clean slate" though. The MVP will rather be to send multiple grids to the same client, for showing popups on top of a buffer and making special windows standalone, rather than modifying the split layout, see #6619 . Also see #2161 for similar feature request.

**Comment by stacksmith** [comment]
@bfredl  - thanks for clarifications.  Multiple grids to same client is definitely useful.

Client handles for colors may be worth it, although I suppose using the palette mode color as an index may be good enough for colors.  I will take a look at the code to see if any of my grand ideas can be easily implemented against neovim's internal grid (when I get a chance...)

A suggestion - consider thinking of this not as a "**here is an interface to display neovim in your GUI frontend**", but instead "**here is an interface for your application to edit text using neovim**".   The difference between the two is very small, but the range of possibilities the latter allows is enormous.  Embedding a good editor into an application - especially easily and quickly -  is an amazing thing to offer (I've been hoping for something like this for decades across many diverse projects).

**Comment by bfredl** [comment]
> "here is an interface for your application to edit text using neovim". 

This is already a discussed usecase, even if not much work has landed on master yet. For instance buffer change notifications #5269 would be useful.

**Comment by vhakulinen** [comment]
| ...I can see how it could have been done in a slightly different way that would make GUI client implementations really simple and pretty much stateless.

I've just finished implementing the `ui-grid` part of own of my UI (using gtk, cairo, pango). Pretty much all the state I need to keep is the cursor, highlight, fg/bg/sp colors, scroll region and the cell metrics of my GUI. Not sure how much more stateless you could go without adding overheat to the protocol - maybe leave all the coloring to the `put` messages, but is there any added benefit? Tho' its probably different thing when you want to use some other than monospaced fonts.

One thing I've noticed from other implementations is that all of them (e.g. the ones I've read, atleast partly) keeps state of the grid where they then draw from. I just draw on the `put`, `clear`, `eol_clear`, `scroll` (tho' those handlers just modifies the canvas and queues the draw - actual drawing is done once gtk is ready for it).

I've yet to dive into the `ui-global` events or the externalized UI options - only done the part `ui-grid` so far. I've also been stalking the UI development for a ~1year or so which might affect the way I see the whole UI API. And yes, the current UI API pretty much assumes that you have a shell/grid where you draw to. Multiple grids PR(s?) looks to change it so that you can have multiple shless/grids - which effectively means that you can have multiple editor windows which you (I assume) can position the way you want to. I might be wrong tho', since I haven't looked too deep into that.

**Comment by stacksmith** [comment]
PUT messages, as implemented, is an array of strings containing one character each.  To add insult to injury, the string "put" is the first element in this array unlike any other part of the protocol (making parsing even more irregular).

Without 'overheating' the protocol, in fact 'cooling it down', a single string along with any highlight information would take up less space and eliminate the need to keep state altogether.  Eliminating pretty much all of the rest of the display-related protocol.   I would even argue for sending messages to blink the cursor.  We are not really concerned with semantics here, such as modes - all we want is a clean display at this level, with a dumb display device just displaying text.

Implementing a client for this kind of a protocol should take a couple of hours at most.

**Comment by justinmk** [comment]
The string-as-array format is discussed in https://github.com/neovim/neovim/issues/6459.

**Comment by bfredl** [comment]
> To add insult to injury, the string "put" is the first element in this array unlike any other part of the protocol

Wrong. This is consistent with the rest of the protocol. It _must_ be, because serialization logic is auto-generated from a specification.

>  a single string along with any highlight information would take up less space and eliminate the need to keep state altogether. 

Allocating cells from a UTF-8 string is highly non-trivial, and definitely requires keeping track of state. It  depends on unicode version, but is in each version also under-specified ("ambigious" chars, something about emojis, etc), and thus dependent on options. By sending pre-segmented cells, we can know for sure that nvim and the gui will have the same idea on what each cell contains, if not exactly how to display it.

> Implementing a client for this kind of a protocol should take a couple of hours at most.

These time estimates are completely arbitrary. Different people will see different models as the "most natural" and simplest to implement model.

**Comment by stacksmith** [comment]
@bfredl:  You can say "wrong" all you want, but facts are facts.  Obviously you haven't had a chance to look at what is on the wire, which is different from your guess about what your messagepack interfaces are doing.  Look at a protocol dump and you will see that EVERY redraw message except put is followed by an array of data.  Put just has strings, and you have to figure out the number of the glyphs by subtracting one from the length of the "put" message.
In fact, things like "bell", which need no data, have an array of 0 elements after.  But put, which desperately needs a separate data area, does not have one.

I don't know who taught you about UTF8, but it is pretty much trivial to decode UTF-8 strings.  Certainly nothing compared to decoding the rest of this mess sometimes referred to as a 'protocol'.  If you don't get that, you have no business writing protocols.

What you are saying is not even wrong, and I don't really wish to argue about it.  If you don't get it, your loss.

**Comment by FSMaxB** [comment]
@stacksmith

> I don't know who taught you about UTF8, but it is pretty much trivial to decode UTF-8 strings. Certainly nothing compared to decoding the rest of this mess sometimes referred to as a 'protocol'. If you don't get that, you have no business writing protocols.

Well not exactly. It is pretty simple to convert a UTF-8 string into a list of codepoints. But it is highly nontrivial to convert these codepoints to actual characters that can be displayed on a screen (if there is some concept of distinguishable characters in the scripture at all).

A few examples:
* "👨" (U+1F468 Man) followed by "🏽" (U+1F3FD Medium skin tone) combine to "👨🏽" (Man: Medium skin tone)
* "e" (U+65) followed by  ́ (U+301) combines to é.
* Even worse: "👩" followed by "🏽" (U+1F3FD Medium skin tone) and "💻" (U+1F4BB Laptop Computer) combines to "👩🏽‍💻" (Woman technologist: Medium skin tone)
* Country flags are combined from an alphabet of letters from a to z (Regional indictor symbol letters) that get combined in pairs to form international language codes. Like "🇺" + "🇸" = "🇺🇸"
* Even more crazy stuff not just emoji.
* Just take a look at the Telugu language.

I recommend you watch [this talk about getting text on a page/screen](https://www.youtube.com/watch?v=Is4PW6f4Pk4)

**Comment by stacksmith** [comment]
@FSMaxB: true enough.  Now, do you think segregating such characters into an array is more helpful than keeping them in a single string?

OT: Skin tone? Yikes.

**Comment by jamessan** [comment]
> If the concern is not knowing the length of the string

Did you look at the referenced [issue](https://github.com/neovim/neovim/issues/6459)?  The conversation there describes some of the reasons behind the current design.

> You can say "wrong" all you want, but facts are facts

You can say "facts" all you want, but that doesn't make them so.

Development is about trade-offs.  We may have chosen some different trade-offs than you would have.  That's fine.  We can iterate from there, but not if you can't have a constructive discussion.

If you're just going to resort to ad hominem and insults, then take your critiques elsewhere.

**Comment by stacksmith** [comment]
@jamessan - OK, if you want to think that the put command in fact has a data array and is just like the rest of the commands, have it your way.

If by constructive discussion you mean others agreeing with your viewpoint (which I am sure has merits from the neovim side, but is extremely arbitrary and painful from the client side) - hey, it's your project.  I am happy to step off.  I certainly am not paid enough to 'critique' your work.

**Comment by bfredl** [comment]
> You can say "wrong" all you want, but facts are facts. Obviously you haven't had a chance to look at what is on the wire, which is different from your guess about what your messagepack interfaces are doing. Look at a protocol dump and you will see that EVERY redraw message except put is followed by an array of data. Put just has strings, and you have to figure out the number of the glyphs by subtracting one from the length of the "put" message.
In fact, things like "bell", which need no data, have an array of 0 elements after. But put, which desperately needs a separate data area, does not have one.

Of course I have looked into the wire format, I rewrote half of the serialization logic (not msgpack itself, but how events gets buffered, batched and flushed as redraw events) some month ago. I don't disagree it is a bit awkward. I just disagreed with the statement I quoted, " the string "put" is the first element in this array unlike any other part of the protocol" , this applies to all  events, regardless if they happen to be coalesced or not (which is in turn not strictly tied to "put", it just happens most frequently). You can't derive a _factual_ disagreement from a disagreement in opinion.

> I don't know who taught you about UTF8, but it is pretty much trivial to decode UTF-8 strings

Which is why I never talked about decoding, only about cell allocation. Even if making one reasonable cell algorithm also is "trivial", making one that works exactly the same as vim in every edge case might not be. Especially as the behavior depend on options....

> you can just write into the text grid

No you can't as "virtual columns" have semantic effect in vim. If the source of truth (in, say, a future protocol that aims to support multi-size text, pictures etc) moves from nvim to the GUI, the gui must report back the cell allocation to vim somehow.

> If by constructive discussion you mean others agreeing with your viewpoint 

You can disagree with someone's viewpoint without making unnecessarily negative personal remarks like "you have no business writing protocols."

**Comment by jamessan** [comment]
Disagreement is how this project started and is how the project progresses.  We're more than welcome to hear disagreement ... as long as you can do that civilly.

> I certainly am not paid enough to 'critique' your work.

That's exactly the point.  We're all volunteering our time, so take a breather.

Describe your ideas so we can see have a discussion about it, but also realize that just because you have a different approach doesn't mean that we'll agree it's right (in part or entirety) for our project.  After all, we're the ones that have to maintain it.

**Comment by stacksmith** [comment]
Listen, I am coming here having implemented a decoder for messagepack (a pretty compact protocol that is pretty close to a stateless representation of a data tree) to reading data from neovim.  I don't think expecting data to be sent in a minimally documented, consistent and efficient manner is a 'wacky idea' that can be disagreed with. 

I don't understand the desire to argue that put is just like the rest the rest of the "redraw" family.  I hand-built a grammar for decoding the "redraw" protocol, and I can tell you that from reading documentation and trying to decode data, that "put" is very peculiar, unless all redraw sub-messages can send multiple data items.  In the wild, all have a single data array; only "highlight_set" and "mode_change" sometimes have two.  This should really be documented.  

There are some good reasons to use messagepack: compactness of representation, state-machine-friendly format (reading a byte at a time in a non-threaded environment), minimal state, etc.  However, none of these are taken advantage of by the neovim protocol.  

The problem here is that the layers of the protocol have not been separated.  The 'redraw' protocol is not the place to fix representation issues.  If you care about correct representation of glyphs, the place to do it is before displaying, by negotiating the glyph issues with the client.   By the time I am displaying glyphs, it is in many cases too late for me, the client, to negotiate representation with the OS or underlying libraries.

It looks to me like the decision of whether this is a dumb-display protocol or a high-level GUI protocol has not been made, so it is an arbitrary subset of both.  As a GUI client, I don't care about mode_change commands - I can't act on them anyway.  What I care about is the color of the glyphs and background, in my representation (which neovim can cache for me instead of me having to guess it from what mode I am in).  I don't care about things like cursor wait and blink times.  If neovim expects me to blink a cursor, it should send a message to turn it on at x-y and later to turn it off.  What I do care about are knowing when the server disconnects due to user typing ":q!"

And again, instead of taking things personally and defending your code or decisions, it would be really nice to step back and look at what your users have to deal with.  As I said before, my choices here are bringing up the problem here, quietly bending over backwards and getting things to work (certainly easier than this), or just not bothering, which is a tempting suggestion.

**Comment by stacksmith** [comment]
It is not an issue of efficiency, but minimalism and correctness.  Keeping unnecessary state is always problematic as it inevitably leads to discrepancies and is a nightmare to maintain.  As I stated before, if the goal here is displaying a neovim screen on some other platform, that should in fact be the goal of the protocol, not "let me send you some data I have in a manner that I find easy to automatically specify"

My reason for implementing a state-machine messagepack decoder is stability in single-threaded environments.  I can read a byte at a time as the situation permits.

I think you may be making assumptions about how messagepack is decoded based on some particular implementation you have.  Messagepack allows you to dispatch on headers before data is in (again, not an optimization but a simplification) or write a byte-matching parser (necessary for state-machine type scenarios - as opposed to read-and-dispatch linear code).  These things are harder to construct if you need to look back to see how many characters are being sent.  There is a precedent for following a string command (which is bad enough) with a data array in many places.  This brings the data size to the parser that knows what to do with it, at the right time, with no state.  Sending a variable-size array with the first item being a string command that dictates how the rest of the data is interpreted is a really bad idea.

My contribution here is to take all the flogging for saying it like it is, in the hope that someone gets what I am saying.  Neovim already has all the information to display text, so it's only a question of how to get it to the client without making it jump through all kinds of hoops.  And I will say again, that implementing a client feels like some kind of a lovecraftian hazing ritual.

I think it's good to understand and be aware of one's limitations whether it's defining protocols or winning popularity contests (this one obviously not my forte).

**Comment by bfredl** [comment]
> I think you may be making assumptions about how messagepack is decoded based on some particular implementation you have. Messagepack allows you to dispatch on headers before data is in (again, not an optimization but a simplification) or write a byte-matching parser (necessary for state-machine type scenarios - as opposed to read-and-dispatch linear code). These things are harder to construct if you need to look back to see how many characters are being sent. 

When the first bytes in a "redraw" event arrive the entire msgpack data has already been composed on the nvim side, even in a byte-level mostly-non-allocating decoder implementation, bailing out and waiting for a few more bytes to arrive is not as expensive as I think you think it is. Also, at no point should there be a need to "look back"; or at least @tarruda:s own libmpack does mostly allocation free decoding without lookback.

> It is not an issue of efficiency, but minimalism and correctness. 

No one here has really disagreed with the claim that the wire format can be improved. But you have to understand if you argue a mosty correct main point with invalid arguments and side points people will point this out. People are not actively trying to be pedants, rather this is part of the natural process of eliminating potential misunderstandings. And so: no matter how "awkward" the wire format is (which I to some extent agree with) or even "inefficient" (again, if this is a big deal depend on where your bottleneck is), it is perfectly self-consistent and possible to treat in a systematic way, for the vast majority of implementations that are not concerned with byte-level optimization.

> let me send you some data I have in a manner that I find easy to automatically specify

The protocol is not inefficient because it is automated, the automation came later. On the contrary, the auto-generation will make it _easier_ to implement a better wire format. because we do not need to implement the better rule for each event, rather just once in the generation logic.

**Comment by stacksmith** [comment]
@bfredl - you are making some assumptions here.  If the UI is running on a different machine over a network, many situations will lock a decoder waiting for the next byte.  In this case, even knowing that the data is not available is not helpful because the decoder is deep inside some loop waiting for the third character of "redraw" string, for instance.  There is just no good way to recover from this.  In a single-threaded situation this is a disaster.  

If the decoder needs to decode the first element of the array to figure out what to do, it needs to keep the array size somewhere, and send it to the code that needs to act on it.  These are on very different levels, and if you ever implement such a decoder yourself, you will see that the code reading the string has the size of the string, while the code reading the array has the size of the array (now stacked above).  You can argue terminology, but the code reading the array now has to somehow communicate its size to the code that was invoked while processing the first element, or the element code has to crawl up the stack to get array size.  This is a level confusion, and is entirely unnecessary.

Again, efficiency has nothing to do with any of it, other than being a side-effect of minimalism and correctness.  Byte-level attention is the result of my requirements of state-machine decoding.

**Comment by bfredl** [comment]
> @bfredl - you are making some assumptions here.

I am. Any well designed protocol is based on assumptions. This one is one of the few that _is_ well stated in the documentation:

> Events must be handled in order. The user should only see the updated screen
state after all events in the same "redraw" batch are processed (not any
intermediate state after processing only part of the array).

Of course, trying to represent to the user the incomplete information available in an incomplete "redraw" event might be a nice addition (if you somehow signal this is an "unofficial" state), but it is not something that is prioritized (we _might_ make this nicer in a ui protocol 2.0, as this hereby _is_ a requested feature, but no promises) nor should you expect these states to look very pretty.

>  while the code reading the array has the size of the array (now stacked above).

Why does the size of the array need to be "stacked above"? Why not store it in the same level as it is being used? Or even modified in place; if we're really speaking byte level efficiency, you don't need to keep the total size and counter, but only the number of _remaining_ objects until you return to the previous level. You can even "fuse" two object layers manually, and only store the total number of objects to decode until some chosen "safe point" , like the end of the current msgpack-rpc message (so you can fast forward to the end of a message, after decode error in multiple levels inside that message), or "redraw" entry list, or whatever. I've been thinking of implementing this design myself in my stale Julia client, to implement REPL integration, but never had the time to do it...

> Byte-level attention is the result of my requirements of state-machine decoding.

A compromise between state machine and multi-threading are co-routines. This way you can keep the basic UI responsible even if the decoder is stuck in the middle of a message, without manually creating a state machine managing every incomplete state. I would use co-routines in Julia, but I don't know about Common Lisp...

**Comment by stacksmith** [comment]
It's not about the order. It's about possible delays on the wire.  The UI should not become unresponsive because of wire delays.

It has to do with being able to implement a generic message decoder that can be customized by plugging in a 'grammar' that runs alongside.  Stacking occurs naturally as the decoder needs to enter and exit aggregate types such as arrays.  As it happens, the most natural way of implementing such a system is to respond to a match when it is found (and not the next state-machine cycle which may never come), and the match is found inside the string decoding code.  

Another way to look at it is that the matching/parsing code should not really be linked to the messagepack this tightly.  It has its own stacking issues as it follows the its parsing graph independent of the messagepack's stack.  It should really look forward only, working on the data it needs to decode, regardless of what dataset it is a part of.  Easily accomplished by sending a command, followed by a data aggregate.  This is just common sense.

So now we have a choice: have all arrays keep the size visible as global state or provide routines for the parser to ask the decoder the size of the array above and subtract one.  Why? because the 'put' command does not provides neither a count of data items to put, nor a container with such data.

And yes, error recovery is another issue, as there is no safe checkpoint to resync to.

**Comment by bfredl** [comment]
> Another way to look at it is that the matching/parsing code should not really be linked to the messagepack this tightly. It has its own stacking issues as it follows the its parsing graph independent of the messagepack's stack. It should really look forward only, working on the data it needs to decode, regardless of what dataset it is a part of. 

All of this is possible already with the current protocol, but it might require a sightly different decoder design than the one you have implemented. In fact with my design of a alloc-free decoder-parser (which I probably have/will fail to describe in words, I hope to put it down in code some time) the  inefficiently of our "put" representation will only cost one extra line of code, a decode-expect of the inner array as a 1-element fixarray, or two lines to allow argument extension, i e first decode-expect an array asking its interior to be pushed, and after consuming the first element pop-fastforward eventual unexpected arguments.

**Comment by stacksmith** [comment]
whatever.

**Comment by bfredl** [comment]
Significant parts of this is fixed in 0.3.2:

- all highlights are pre-allocated
- A single event is used to update the changed part of each line
- As a result event interdependence is greatly reduced. Cursor event is only used for visible cursor position.
